File: DotNet/Controllers/ModelController.cs
using Microsoft.AspNetCore.Mvc;
using System.Collections.Generic;
using System.Threading.Tasks;
using Vehicle_Configurator_Project.IServices;
using Vehicle_Configurator_Project.Models;

namespace Vehicle_Configurator_Project.Controllers
{
    [ApiController]
    [Route("api/models")]
    public class ModelController : ControllerBase
    {
        private readonly IModelService _modelService;
        private readonly ISegmentService _segmentService;
        private readonly IManufacturerService _manufacturerService;
        private readonly IVehicleDetailService _vehicleDetailService;

        public ModelController(
            IModelService modelService,
            ISegmentService segmentService,
            IManufacturerService manufacturerService,
            IVehicleDetailService vehicleDetailService)
        {
            _modelService = modelService;
            _segmentService = segmentService;
            _manufacturerService = manufacturerService;
            _vehicleDetailService = vehicleDetailService;
        }

        [HttpGet]
        public async Task<ActionResult<List<Model>>> GetAllModels()
        {
            var models = await _modelService.GetAllModelsAsync();
            return Ok(models);
        }

        [HttpPost]
        public async Task<ActionResult<Model>> AddModel([FromBody] Model model)
        {
            var created = await _modelService.CreateModelAsync(model);
            return CreatedAtAction(nameof(GetModelById), new { modelId = created.ModelId }, created);
        }

        [HttpGet("{modelId:int}")]
        public async Task<ActionResult<Model>> GetModelById(int modelId)
        {
            var model = await _modelService.GetModelByIdAsync(modelId);
            if (model == null) return NotFound();
            return Ok(model);
        }

        [HttpGet("by-segment/{segmentId:int}")]
        public async Task<ActionResult<List<Model>>> GetModelsBySegment(int segmentId)
        {
            var segment = await _segmentService.GetSegmentByIdAsync(segmentId);
            if (segment == null) return NotFound();

            var models = await _modelService.GetModelsBySegmentAsync(segment);
            return Ok(models);
        }

        [HttpGet("by-segment/{segmentId:int}/manufacturer/{manufacturerId:int}")]
        public async Task<ActionResult<List<Model>>> GetModelsBySegmentAndManufacturer(int segmentId, int manufacturerId)
        {
            var segment = await _segmentService.GetSegmentByIdAsync(segmentId);
            if (segment == null) return NotFound();

            var manufacturer = await _manufacturerService.GetManufacturerByIdAsync(manufacturerId);
            if (manufacturer == null) return NotFound();

            var models = await _modelService.GetModelsByManufacturerAndSegmentAsync(manufacturer, segment);
            return Ok(models);
        }

        [HttpGet("default/{modelId:int}")]
        public async Task<ActionResult<Model>> GetModelWithDefaults(int modelId)
        {
            var model = await _modelService.GetModelWithDefaultComponentsAsync(modelId);
            if (model == null) return NotFound();
            return Ok(model);
        }

        [HttpGet("configurable/{modelId:int}")]
        public async Task<ActionResult<List<VehicleDetail>>> GetConfigurableComponents(int modelId)
        {
            // Note the char 'Y' instead of string "Y" to avoid type errors
            var components = await _vehicleDetailService.GetByModelIdAndIsConfigurableAsync(modelId, 'Y');
            return Ok(components);
        }

        [HttpGet("components/{modelId:int}")]
        public async Task<ActionResult<List<VehicleDetail>>> GetAllComponents(int modelId)
        {
            var components = await _modelService.GetComponentsByModelIdAsync(modelId);
            return Ok(components);
        }

        [HttpGet("alternate-components/{modelId:int}")]
        public async Task<ActionResult<Dictionary<char, Dictionary<int, List<AlternateComponent>>>>> GetAlternateComponentsGrouped(int modelId)
        {
            var grouped = await _modelService.GetGroupedAlternatesByModelAsync(modelId);
            return Ok(grouped);
        }

        [HttpGet("price/{modelId:int}")]
        public async Task<ActionResult<decimal>> GetBasePrice(int modelId)
        {
            var price = await _modelService.GetBasePriceByModelIdAsync(modelId);
            if (price == null) return NotFound();
            return Ok(price);
        }
    }
}

File: JAVA/VehicleConfigurator/src/main/java/com/example/controllers/ModelController.java
package com.example.controllers;
import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import com.example.entities.AlternateComponent;
import com.example.entities.Manufacturer;
import com.example.entities.Model;
import com.example.entities.Segment;
import com.example.entities.VehicleDetail;
import com.example.repositories.AlternateComponentRepository;
import com.example.services.ManufacturerService;
import com.example.services.ModelService;
import com.example.services.SegmentService;
import com.example.services.VehicleDetailService;

@RestController
@RequestMapping("/models")
public class ModelController {

    @Autowired
    private ModelService modelService;
    
    @Autowired
    private SegmentService segmentService;
    
    @Autowired
    private ManufacturerService manufacturerService;
    
    @Autowired
    private VehicleDetailService vehicleDetailService;
    
    
    @Autowired
    public ModelController(ModelService modelService) {
        this.modelService = modelService;
    }

    @GetMapping
    public List<Model> getAllModels() {
        return modelService.getAllModels();
    }

    @PostMapping
    public Model addModel(@RequestBody Model model) {
        return modelService.createModel(model);
    }
    
    @GetMapping("/{modelId}")
    public ResponseEntity<Model> getModelById(@PathVariable Integer modelId) {
        Model model = modelService.getModelById(modelId);
        return ResponseEntity.ok(model);
    }
    
    @GetMapping("/by-segment/{segmentId}")
    public List<Model> getModelsBySegment(@PathVariable int segmentId) {
        Segment segment = segmentService.getSegmentById(segmentId);
        return modelService.getModelsBySegment(segment); 
    }
    
    @GetMapping("/by-segment/{segmentId}/manufacturer/{manufacturerId}")
    public List<Model> getModelsBySegmentAndManufacturer(@PathVariable int segmentId,
                                                        @PathVariable int manufacturerId) {
        Segment segment = segmentService.getSegmentById(segmentId);
        Manufacturer manufacturer = manufacturerService.getManufacturerById(manufacturerId);
        return modelService.getModelsByManufacturerAndSegment(manufacturer, segment);
    }
    
    //previous
    @GetMapping("/default/{modelId}")
    public ResponseEntity<Model> getModelWithDefaults(@PathVariable Integer modelId) {
        Model model = modelService.getModelWithDefaultComponents(modelId);
        return ResponseEntity.ok(model);
    }
    
    //new
//    @GetMapping("/default/{modelId}")
//    public ResponseEntity<?> getModelDefaultConfig(@PathVariable int modelId) {
//        Model model = modelService.getDefaultConfigWithCoreComponents(modelId);
//        if (model == null) return ResponseEntity.notFound().build();
//        return ResponseEntity.ok(model);
//    }
    
    @GetMapping("/configurable/{modelId}")
    public ResponseEntity<List<VehicleDetail>> getConfigurableComponents(@PathVariable int modelId) {
        List<VehicleDetail> components = vehicleDetailService.getByModelIdAndIsConfigurable(modelId, 'Y');
        return ResponseEntity.ok(components);
    }
    
    @GetMapping("/components/{modelId}")
    public ResponseEntity<List<VehicleDetail>> getAllComponents(@PathVariable int modelId) {
        List<VehicleDetail> components = modelService.getComponentsByModelId(modelId);
        return ResponseEntity.ok(components);
    }
    
    @GetMapping("/alternate-components/{modelId}")
    public ResponseEntity<Map<Character, Map<Integer, List<AlternateComponent>>>> getAlternateComponentsGrouped(
            @PathVariable int modelId) {
        Map<Character, Map<Integer, List<AlternateComponent>>> grouped = modelService.getGroupedAlternatesByModel(modelId);
        return ResponseEntity.ok(grouped);
    }
    
    @GetMapping("/price/{modelId}")
    public ResponseEntity<BigDecimal> getBasePrice(@PathVariable Integer modelId) {
        BigDecimal price = modelService.getBasePriceByModelId(modelId);
        if (price == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(price);
    }
    
}
